qiskit/pulse/pulse.py

```diff
@@ -0,0 +1,164 @@
+"""
+
+import ast
+import cmath
+import inspect
+
+import numpy as np
+from matplotlib import get_backend as get_matplotlib_backend, pyplot as plt
+from scipy.interpolate import CubicSpline
+
+from qiskit.exceptions import QiskitError
+
+MATH = ['exp', 'log', 'log10', 'sqrt', 'acos', 'asin', 'atan', 'cos', 'sin', 'tan',
+        'acosh', 'asinh', 'atanh', 'cosh', 'sinh', 'tanh', 'pi', 'e', 'tau']
+
+
+class Pulse:
+    """Pulse specification."""
+
+    def __init__(self, envfunc, t_gate, t_intv, **kwargs):
+        """Create a new pulse envelope.
+        Args:
+            envfunc (str): equation describing pulse envelope
+            t_gate (float): gate time
+            t_intv (float): time interval of sampling points
+            **kwargs: initial values of pulse parameter
+        Raises:
+            QiskitError: when pulse envelope is not in the correct format.
+        """
+        self._param = {}
+        self._ast = None
+        self.t_intv = t_intv
+        self.t_gate = t_gate
+
+        if isinstance(envfunc, str):
+            self._ast, pars = Pulse.eval_expr(envfunc)
+            self._param = {par: kwargs.get(par, 0) for par in pars}
+        else:
+            raise QiskitError('Incorrect specification of pulse envelope function.')
+
+    @property
+    def param(self):
+        """Get parameters for describing pulse envelope
+
+        Returns:
+            dict: pulse parameters
+        """
+        return self._param
+
+    @param.setter
+    def param(self, param_new):
+        """Set parameters for describing pulse envelope
+
+        Args:
+            param_new (dict): dictionary of parameters
+        Raises:
+            QiskitError: when pulse parameter is not in the correct format.
+        """
+        if isinstance(param_new, dict):
+            for key, val in self._param.items():
+                self._param[key] = param_new.get(key, val)
+        else:
+            raise QiskitError('Pulse parameter should be dictionary.')
+
+    def tolist(self):
+        """Calculate and output pulse envelope as a list of complex values [re, im]
+
+        Returns:
+            list: complex pulse envelope at each sampling point
+        Raises:
+            QiskitError: when envelope function is not initialized.
+        """
+        if self._ast:
+            math_funcs = dict(inspect.getmembers(cmath))
+            _dict = {k: math_funcs.get(k, None) for k in MATH}
+            _dict.update(self._param)
+            envelope = []
+
+            _time = 0
+            while _time <= self.t_gate:
+                _dict['t'] = _time
+                value = eval(compile(self._ast, filename='', mode='eval'),
+                             {"__builtins__": None}, _dict)
+                envelope.append([np.real(value), np.imag(value)])
+                _time += self.t_intv
+        else:
+            raise QiskitError('Envelope function is not defined.')
+
+        return envelope
+
+    def plot(self, nop=1000):
```

Comment:
Can we separate the plotting from this module, and find a place for it in `qiskit/tools/visualization` or similar? Both for modularity and for practical concerns, as `matplotplib` is not a terra dependency - [only optional](https://github.com/Qiskit/qiskit-terra/blob/master/setup.py#L55).